rasa_sdk/endpoint.py

```diff
@@ -4,23 +4,280 @@
+            async with session.request(
+                method,
+                url,
+                headers=headers,
+                params=self.combine_parameters(kwargs),
+                ssl=sslcontext,
+                **kwargs,
+            ) as response:
+                if response.status >= 400:
+                    raise ClientResponseError(
+                        response.status, response.reason, await response.content.read()
+                    )
+                try:
+                    return await response.json()
+                except ContentTypeError:
+                    return None
+
+    @classmethod
+    def from_dict(cls, data: Dict[Text, Any]) -> "EndpointConfig":
+        return EndpointConfig(**data)
+
+    def copy(self) -> "EndpointConfig":
+        return EndpointConfig(
+            self.url,
+            self.params,
+            self.headers,
+            self.basic_auth,
+            self.token,
+            self.token_name,
+            **self.kwargs,
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(self, type(other)):
+            return (
+                other.url == self.url
+                and other.params == self.params
+                and other.headers == self.headers
+                and other.basic_auth == self.basic_auth
+                and other.token == self.token
+                and other.token_name == self.token_name
+            )
+        else:
+            return False
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
+
+
+# copied from rasa.utils.endpoints
+class ClientResponseError(aiohttp.ClientError):
+    def __init__(self, status: int, message: Text, text: Text) -> None:
+        self.status = status
+        self.message = message
+        self.text = text
+        super().__init__(f"{status}, {message}, body='{text}'")
 
+
+# copied from rasa.utils.endpoints
+def concat_url(base: Text, subpath: Optional[Text]) -> Text:
+    """Append a subpath to a base url.
+
+    Strips leading slashes from the subpath if necessary. This behaves
+    differently than `urlparse.urljoin` and will not treat the subpath
+    as a base url if it starts with `/` but will always append it to the
+    `base`.
+
+    Args:
+        base: Base URL.
+        subpath: Optional path to append to the base URL.
+
+    Returns:
+        Concatenated URL with base and subpath.
+    """
+    if not subpath:
+        if base.endswith("/"):
+            logger.debug(
+                f"The URL '{base}' has a trailing slash. Please make sure the "
+                f"target server supports trailing slashes for this endpoint."
+            )
+        return base
+
+    url = base
+    if not base.endswith("/"):
+        url += "/"
+    if subpath.startswith("/"):
+        subpath = subpath[1:]
+    return url + subpath
+
+
+local_domain = dict()
+domain_endpoint = None
+
+logger.debug(f"Domain Endpoint is {DOMAIN_ENDPOINT}")
+
+if DOMAIN_ENDPOINT:
```

Comment:
i am a little worried about the service dependency this introduces. with the change added, a rasa SDK deployment now depends on rasa open source. previously, only rasa open source dependent on the SDK. with this change the dependency becomes bidirectional where there was previously only a unidirectional dependency.

Reply:
But only if you enable this feature. If you don't - or you have another compatible domain endpoint somewhere else, you don't have the dependency.

Reply:
Also, it never tries to contact the domain endpoint if the Rasa server sends a domain with the request (default behavior), so you need to actively switch that off in Rasa server with the env var introduced in [Rasa PR11966](https://github.com/RasaHQ/rasa/pull/11966)